### PR TITLE
Fix RemoteTaskRunner's auto-scaling

### PR DESCRIPTION
**Pull request by leventov**
**Problem**
Overlord leadership blink causes shutting down `RemoteTaskRunner` and creating a new `RemoteTaskRunner`. But the second `RemoteTaskRunner` couldn't start, because they share global `ResourceManagementStrategy`, which shuts down it's executor when the first `RemoteTaskRunner` calls `resourceManagementStrategy.stopManagement()`.

**How it is fixed**
`ResourceManagementStrategy` is now an immutable object, it creates mutable stateful `ProvisioningService` objects upon request for each individual `RemoteTaskRunner`. Each `ProvisioningService` has it's own executor, which is shut down from `RemoteTaskRunner.stop()`.

**Testing**
Added `OverlordBlinkLeadershipTest`, demonstrating this issue.

**What else is changed**
`ResourceManagementStrategy` is refactored, first of all it is renamed to `ProvisioningStrategy` (and all related classes are renamed as well), because `Provisioning` is a little more distinctive keyword within Druid codebase than plain `ResourceManagement`.

Also `AbstractWorkerProvisioningStrategy` (former `AbstractWorkerResourceManagementStrategy`) is refactored towards composition from inheritance, `Provisioner` interface is extracted.

**Size of this PR**
Conceptual size of this PR is smaller than you might think looking at +710 -528, e. g. in `PendingTaskBasedWorkerProvisioningStrategy` and `SimpleWorkerProvisioningStrategy.java` mostly just code indentation has changed, that generated many "changed lines".

### CODE REVIEW

**Comment by drcrallen** [comment]
Semi related: https://github.com/druid-io/druid/issues/3498 https://github.com/druid-io/druid/issues/3428

**Comment by drcrallen** [comment]
https://github.com/druid-io/druid/issues/2294 also semi-related.

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/PendingTaskBasedWorkerProvisioningStrategy.java [hunk 1]
```diff
@@ -53,56 +52,74 @@
...
-  public boolean doProvision(WorkerTaskRunner runner)
+  public Provisioner makeProvisioner(WorkerTaskRunner runner)
   {
-    Collection<Task> pendingTasks = runner.getPendingTaskPayloads();
-    Collection<ImmutableWorkerInfo> workers = runner.getWorkers();
-    synchronized (lock) {
+    return new PendingProvisioner(runner);
+  }
+
+  private class PendingProvisioner implements Provisioner
```
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/PendingTaskBasedWorkerProvisioningStrategy.java]: Explored this but didn't find anything that could be extracted as a method with a sensible name and "abstraction", so didn't extract

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/PendingTaskBasedWorkerProvisioningStrategy.java [hunk 2]
```diff
@@ -53,56 +52,74 @@
...
         workerConfigRef,
-        resourceManagementSchedulerConfig,
-        factory.create(1, "PendingTaskBasedResourceManagement-manager--%d")
+        provisioningSchedulerConfig,
+        new Supplier<ScheduledExecutorService>()
+        {
+          @Override
+          public ScheduledExecutorService get()
+          {
+            return ScheduledExecutors.fixed(1, "PendingTaskBasedResourceManagement-manager--%d");
```
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/PendingTaskBasedWorkerProvisioningStrategy.java]: Fixed

**Comment by leventov** [comment]
Could anybody please review this?

**Comment by jihoonson** [comment]
I'll review tonight.

File: indexing-service/src/test/java/io/druid/indexing/overlord/OverlordBlinkLeadershipTest.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  /**
+   * Test that we can start taskRunner, then stop it (emulating "losing leadership", see {@link
+   * TaskMaster#stopLeading()}), then creating a new taskRunner from {@link
+   * org.apache.curator.framework.recipes.leader.LeaderSelectorListener#takeLeadership} implementation in
+   * {@link TaskMaster} and start it again.
+   */
+  @Test(timeout = 10_000)
+  public void testOverlordBlinkLeadership()
```
Comment (leventov) [inline indexing-service/src/test/java/io/druid/indexing/overlord/OverlordBlinkLeadershipTest.java]: @jihoonson thanks for review!

The logic of `AbstractWorkerResourceManagementStrategy` (now `AbstractWorkerProvisioningStrategy`) is simplified and it doesn't use LifecycleLock anymore. It also creates `ProvisioningService` in `RemoteTaskRunner.start()`, that is closed in `RemoteTaskRunner.stop()`. All those methods are called from a single thread in this test. So I don't think any change to this test is needed.

Comment (jihoonson) [inline indexing-service/src/test/java/io/druid/indexing/overlord/OverlordBlinkLeadershipTest.java]: Hmm, I was wondering how I can reproduce the problem using this test in the master branch. But, it seems not important. The test looks good to me now.

**Comment by leventov** [comment]
@jihoonson thanks for review. Could you please merge it?

**Comment by jihoonson** [comment]
Sure.

**Comment by leventov** [comment]
@jihoonson thank you!

**Comment by jihoonson** [comment]
@leventov no problem!

**Pull request merged**